runtime/lua/vim/pack/_lsp.lua

```diff
@@ -22,6 +23,48 @@ local get_confirm_bufnr = function(uri)
   return tonumber(uri:match('^nvim%-pack://(%d+)/confirm%-update$'))
 end
 
+local group_header_pattern = '^# (%S+)'
+local plugin_header_pattern = '^## (.+)$'
+
+--- @return { group: string?, name: string?, from: integer?, to: integer? }
+local get_plug_data_at_lnum = function(bufnr, lnum)
+  local lines = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
+  --- @type string, string, integer, integer
+  local group, name, from, to
+  for i = lnum, 1, -1 do
+    group = group or lines[i]:match(group_header_pattern) --[[@as string]]
```

Comment:
Does `vim.regex:match_line()` avoid a lot of this code? It can match a buffer line.

(We should add `nvim_buf_match()` but oh well, vim regex is fine :)

Reply:
I can't see now how it can remove *a lot* of code.

To get a full structural data about particular line number, it has to go up to see its nearest headings (second level for plugin name, first level for plugin group like "Error"/"Update"/"Same") and down to see where it ends.

`vim.regex:match_line()` would still require iterating over lines, the same as this approach. The only code save I see is that there would be no need to get all buffer lines.

The patterns for the name and group headings are also used later, so they are also not remove-able by refactoring this part.